### ISSUE TITLE
False negative for `Style/RedundantRegexpArgument` and `start_with?`/`end_with?`

### ISSUE THREAD

**Issue by Earlopain** [rubocop/rubocop#13246]
`start_with?` and `end_with?` accept multiple arguments, so something like `start_with?(/foo|bar/)` can be replaced by `start_with?("foo, "bar")`. Simple benchmarking shows this to be ~1.75x-2.25x as fast, depending on how many elements in the array potentially match.

I read https://github.com/rubocop/rubocop/pull/11595 and the reason why it is not in `Performance` is because it is has a positive effect on readability.
RuboCop has code like this:

```rb
p.start_with?(/docs|Gemfile|README|CHANGELOG/)
# which would turn into 
p.start_with?('docs', 'Gemfile', 'README', 'CHANGELOG')
```

Which I still think is still fine for this style cop? If not, please move this issue to `rubocop-performance` repo.

Quick and dirty benchmark:

```rb
# frozen_string_literal: true

require 'benchmark/ips'

ARRAY = %w[foo_aionnfia oinianfi oiabnf aowinfaw oinawf foo_aoisd bar_ianffna oanf aowfnwf].freeze
REGEXP = /foo_|bar_/.freeze

Benchmark.ips do |x|
  x.report('str') do
    ARRAY.each { _1.start_with?('foo_', 'bar_') }
  end
  x.report('inline regexp') do
    ARRAY.each { _1.start_with?(/foo_|bar_/) }
  end
  x.report('const regexp') do
    ARRAY.each { _1.start_with?(REGEXP) }
  end
  x.compare!
end

ruby 3.3.5 (2024-09-03 revision ef084cc8f4) [x86_64-linux]
Warming up --------------------------------------
                 str   158.362k i/100ms
       inline regexp    71.410k i/100ms
        const regexp    71.320k i/100ms
Calculating -------------------------------------
                 str      1.579M (± 0.4%) i/s  (633.34 ns/i) -      7.918M in   5.014912s
       inline regexp    710.794k (± 0.5%) i/s    (1.41 μs/i) -      3.570M in   5.023359s
        const regexp    713.881k (± 0.6%) i/s    (1.40 μs/i) -      3.637M in   5.095303s

Comparison:
                 str:  1578936.6 i/s
        const regexp:   713881.4 i/s - 2.21x  slower
       inline regexp:   710794.4 i/s - 2.22x  slower
```

**Comment by koic** [comment]
As far as I also have researched past, whether multiple strings or a single regular expression is faster depends on the conditions and the target data. In other words, strings are not necessarily always faster. Additionally, I'm doubtful whether converting a single regular expression into multiple strings makes sense, so replacing such a regular expression with multiple strings feels like overkill to me.

```console
$ cat bench.rb
require 'benchmark/ips'

ARRAY = %w[foo_aionnfia oinianfi oiabnf aowinfaw oinawf baz_aoisd bar_ianffna oanf aowfnwf].freeze
REGEXP = /foo_|bar_|baz_|qux_/

Benchmark.ips do |x|
  x.report('str') do
    ARRAY.each { _1.start_with?('foo_', 'bar_', 'baz_', 'qux_') }
  end
  x.report('inline regexp') do
    ARRAY.each { _1.start_with?(/foo_|bar_|baz_|qux_/) }
  end
  x.report('const regexp') do
    ARRAY.each { _1.start_with?(REGEXP) }
  end
  x.compare!
end

$ ruby bench.rb
ruby 3.3.5 (2024-09-03 revision ef084cc8f4) [x86_64-darwin23]
Warming up --------------------------------------
                 str    36.871k i/100ms
       inline regexp    38.008k i/100ms
        const regexp    41.411k i/100ms
Calculating -------------------------------------
                 str    358.247k (± 2.3%) i/s -      1.807M in   5.045819s
       inline regexp    406.760k (± 4.4%) i/s -      2.052M in   5.056361s
        const regexp    397.327k (± 5.8%) i/s -      1.988M in   5.021082s

Comparison:
       inline regexp:   406759.6 i/s
        const regexp:   397327.5 i/s - same-ish: difference falls within error
                 str:   358247.0 i/s - 1.14x  slower
```

**Comment by Earlopain** [comment]
Thanks for that benchmark, I didn't realize you already considered this. Since it depends on a variety of factors, I'll close.

> Additionally, I'm doubtful whether converting a single regular expression into multiple strings makes sense, so replacing such a regular expression with multiple strings feels like overkill to me.

I depends :tm:, if called in a hotspot this can have a unexpectedly large impact compared to the other approach. But that's what profiling tools are for.
